### PR TITLE
Add FFT QSP 

### PR DESCRIPTION
**Pull request by Epsilon1024**
Added fft_qsp module to address [Issue 1076](https://github.com/quantumlib/Qualtran/issues/1076)

This fixes the error issues found by fast_complementary_polynomial by using the algorithm described in [Berntson and Sunderhauf (2024)](https://arxiv.org/abs/2406.04246) (Algorithm 2 based off the code in Figure 1). This method takes the target precision as an input parameter and automatically determines the number of FFT modes needed to guarantee the given precision.

### CODE REVIEW

**Comment by tanujkhattar** [comment]
cc @christoph-riverlane for visibility, who is one of the authors of the linked paper.

**Comment by Epsilon1024** [comment]
I just noticed that N must be even (corollary 2.2).  Let me fix that real quick.

**Comment by anurudhp** [comment]
@Epsilon1024 could you rebase on main and squash all the commits so far? there's a lot of legacy commits, it's hard to keep track of new changes to review.

**Comment by anurudhp** [comment]
Right that's true. I meant more for reviewing incremental changes, but I guess I can just look from the last two commits, so no issue.

**Comment by christoph-riverlane** [comment]
Hi all,
thanks Tanuj for tagging me, and thanks for the interest in our algorithm and paper. Two notes:

* Our proofs require N even. However, the algorithm appears to work also for odd N. (In fact, for our numerical results, we have used both odd and even values of N. We'll add a comment on this to the paper.) I think our proofs can also be extended to odd N.

* The values for N from the proof are not tight: We found that sometimes in practice N can be sufficient that are significantly smaller than the proven bounds, see eg the values of N in Fig 2. Therefore, we had envisioned that in practice, a search procedure might be used, in which the algorithm is run repeatedly with doubled N until the desired accuracy is reached. (We'll add a comment on this to the paper)

Cheers, Christoph

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    verify_generalized_qsp,
+)
+
+
+@pytest.mark.parametrize("degree, tolerance", [(4, 1e-3), (5, 1e-3)])
+def test_complementary_polynomial_quick(degree: int, tolerance: float):
+    random_state = np.random.RandomState(42)
+    for _ in range(2):
+        P = random_qsp_polynomial(degree, random_state=random_state)
+        Q = fft_complementary_polynomial(P, tolerance=tolerance)
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:33]: We don't have to duplicate code for the fast and slow tests, we can instead use `pytest.param` to mark some of the parameters slow. See https://github.com/quantumlib/Qualtran/blob/cc4e3bcf1729cb31dcaa22591610ebb552e82d2b/qualtran/bloqs/qsp/generalized_qsp_test.py#L56-L74 for an example.

Also, only mark tests slow if they take > 1s.

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+)
+def test_real_polynomial_has_real_complementary_polynomial(
+    degree: int, num_tests: int, tolerance: float
+):
+    random_state = np.random.RandomState(42)
+    for _ in range(num_tests):
+        P = random_qsp_polynomial(degree, random_state=random_state, only_real_coeffs=True)
+        Q = fft_complementary_polynomial(P, tolerance=tolerance)
+        Q = np.around(Q, decimals=8)
+        assert np.isreal(Q).all()
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:96]:
```suggestion
        np.testing.assert_allclose(np.imag(Q), 0, atol=tolerance)
```

This should also improve the values of tolerance (this was the case in `fast_qsp_test.py`). Could you check what the smallest tolerance values are such that the tests pass?

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            P, Q, random_state=random_state, rtol=tolerance
+        )
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize("bitsize", [1, 2, 3])
+@pytest.mark.parametrize(
+    "degree, negative_power, tolerance",
+    [(2, 0, 1e-3), (2, 1, 1e-3), (2, 2, 1e-3), (5, 0, 1e-2), (5, 1, 1e-3), (5, 2, 1e-3)],
+)
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:107]: factor out `negative_power` as a separate parameter (as it is uncorrelated with the rest)

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    modes[0] *= 1 / 2  # Note modes are ordered differently in the text
+    modes[N // 2 + 1 :] = 0
+    return modes
+
+
+def fft_complementary_polynomial(
+    P: Union[Sequence[float], Sequence[complex]], tolerance: float = 1e-4
+):
+    """
+    Computes the Q polynomial given P
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:60]:
```suggestion
    """Computes the Q polynomial given P
```

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    # Scale P
+    P = np.array(P)
+    scaled_P = (1 - tolerance / 4) * P
+
+    d = P.shape[0]
+    N = _get_N(tolerance, d)
+
+    # Pad P to FFT dimension N
+    padded_poly = lambda x: np.pad(scaled_P, (0, N - 1))
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:84]: argument `x` is not used?

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    # Evaluate P(omega) at roots of unity omega
+    p_eval = lambda x: np.fft.ifft(padded_poly(x), norm="forward")
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:86]: is the lambda needed? (same for the lambdas below). If not remove them, and if yes, replace it with `def ...`

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import pytest
+
+from qualtran.bloqs.for_testing.matrix_gate import MatrixGate
+
+from .fft_qsp import fft_complementary_polynomial
+from .generalized_qsp_test import (
+    check_polynomial_pair_on_random_points_on_unit_circle,
+    random_qsp_polynomial,
+    verify_generalized_qsp,
+)
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:25]: please update these imports, and use absolute import paths.

ref. https://github.com/quantumlib/Qualtran/blob/main/qualtran/linalg/polynomial/qsp_testing.py

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+def test_complementary_polynomial(degree: int, num_tests: int, tolerance: float):
+    random_state = np.random.RandomState(42)
+
+    for _ in range(num_tests):
+        P = random_qsp_polynomial(degree, random_state=random_state)
+        Q = fft_complementary_polynomial(P, tolerance)
+        check_polynomial_pair_on_random_points_on_unit_circle(
+            P, Q, random_state=random_state, rtol=tolerance
+        )
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:65]: Could you investigate what the best tolerance values are. 

For a comparison, see the updated values in https://github.com/quantumlib/Qualtran/blob/cc4e3bcf1729cb31dcaa22591610ebb552e82d2b/qualtran/bloqs/qsp/fast_qsp_test.py#L27-L45

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    [(2, 3, 1e-4), (3, 3, 1e-3), (4, 1, 1e-3), (5, 1, 1e-2), (10, 3, 1e-2)],
+)
+def test_fast_qsp_on_random_unitaries(degree: int, num_tests: int, tolerance: float):
+    random_state = np.random.RandomState(102)
+
+    for _ in range(num_tests):
+        P = random_qsp_polynomial(degree, random_state=random_state)
+        U = MatrixGate.random(2, random_state=random_state)
+        Q = fft_complementary_polynomial(P, tolerance=tolerance)
+        verify_generalized_qsp(U, P, Q=Q, tolerance=tolerance)
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:80]: ditto (investigate best tolerance values)

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import math
+from typing import Sequence, Union
+
+import numpy as np
+
+
+def _get_r(delta: float, d: int) -> float:
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:20]: briefly describe what this computes, and add references to the paper where this is described.

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    return (1 / (1 - delta)) ** (1 / d)
+
+
+def _get_N0(epsilon: float, delta: float, d: int) -> int:
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:24]: ditto (description + ref. to paper)

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    r = _get_r(delta, d)
+    first_term = 2 / (np.log(r))
+
+    second_term_num = 8 * np.log(1 / delta)
+    second_term_den = epsilon * (r - 1)
+
+    return math.ceil(first_term * np.log(second_term_num / second_term_den))
+
+
+def _get_N(epsilon0: float, d: int) -> int:
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:34]: ditto (description + ref. to paper)

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    epsilon = epsilon0 / 4
+    delta = epsilon0 / (5 * (d + 1))
+
+    result = _get_N0(epsilon, delta, d)
+    if result % 2 == 0:
+        return result
+    return result + 1
+
+
+def _get_scale_factor(epsilon: float) -> float:
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:45]: ditto (description + ref. to paper)

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+    return 1 - (epsilon / 4)
+
+
+def _get_modes(the_log: np.ndarray, N: int) -> np.ndarray:
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:49]: ditto (description + ref. to paper)

**Review by anurudhp — CHANGES_REQUESTED** [review]
Very nice. Left a first round of comments.

Could you investigate what the best tolerance values are for the current implementation? I suspect they are much better than the current test parameters.

File: Untitled.ipynb
Comment (anurudhp) [inline Untitled.ipynb:1]: this notebook seems quite useful, perhaps name it appropriately and move it to `qualtran/bloqs/qsp`?

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+        Q = fft_complementary_polynomial(P, tolerance=tolerance, num_modes=num_modes)
+        np.testing.assert_allclose(np.imag(Q), 0, atol=tolerance)
+        check_gqsp_polynomial_pair_on_random_points_on_unit_circle(
+            P, Q, random_state=random_state, rtol=tolerance
+        )
+
+
+@pytest.mark.parametrize("degree", [2, 3, 4, 5, 10])
+@pytest.mark.parametrize("bitsize", [1, 2, 3])
+def test_fft_qsp_on_random_unitaries(degree: int, bitsize: int):
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:60]: this seems identical to the next test?

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    random_state = np.random.RandomState(102)
+    num_modes = 500
+    tolerance = 1e-6
+    for _ in range(10):
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:64]:
```suggestion
    for _ in range(5):
```

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+
+@pytest.mark.parametrize(
+    "degree, num_modes",
+    [(3, 500), (4, 500), (5, 500), (10, 500), (40, 500), (100, 500), (1000, 5000), (10000, 50000)],
+)
+def test_real_polynomial_has_real_complementary_polynomial(degree: int, num_modes: int):
+    random_state = np.random.RandomState(42)
+    tolerance = 1e-9
+    for _ in range(10):
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:49]:
```suggestion
    for _ in range(5):
```

File: qualtran/bloqs/qsp/fft_qsp_test.py
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:1]: also remember to mark any slow tests (> 1s), if there are any.

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    def calculate_coeff(poly: np.ndarray) -> np.ndarray:
+        """Compute coefficients of Q
+
+        This runs the entire process calling the other intermediate methods.
+
+        Args:
+            poly: The input polynomial, P.
+        Returns:
+            The complementary polynomial, Q.
+        """
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:81]: in the spirit of the other docstrings, I think this should only describe what this step does (i.e. `fft(exp(p))`)

**Review by anurudhp — APPROVED** [review]
looks good! left some comments on docstring fixes, otherwise lgtm

cc @tanujkhattar

File: qualtran/bloqs/qsp/fft_qsp_test.py (hunk 15 above)
Comment (Epsilon1024) [reply qualtran/bloqs/qsp/fft_qsp_test.py:60]: Oh!  You're right!

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+        P = random_qsp_polynomial(degree, random_state=random_state)
+        U = MatrixGate.random(bitsize=bitsize, random_state=random_state)
+        Q = fft_complementary_polynomial(P, tolerance=tolerance, num_modes=num_modes)
+        verify_generalized_qsp(U, P, Q=Q, tolerance=tolerance)
+
+
+@pytest.mark.parametrize("bitsize", [1, 2, 3])
+@pytest.mark.parametrize("degree", [3, 5, 10])
+@pytest.mark.parametrize("negative_power", [0, 1, 2])
```
[the remark this answers is not in the document]
Comment (Epsilon1024) [reply qualtran/bloqs/qsp/fft_qsp_test.py:73]: The bitsize parameter?

File: qualtran/bloqs/qsp/fft_qsp_test.py
Comment (Epsilon1024) [reply qualtran/bloqs/qsp/fft_qsp_test.py:1]: All tests run in < 1s !

File: qualtran/bloqs/qsp/fft_qsp_test.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/qsp/fft_qsp_test.py:73]: no, I meant `negative_power`. It just implements `U^{negative_power} P(U)`, so the `U^{negative_power}` part doesn't depend on `P, Q`

File: Untitled.ipynb
Comment (Epsilon1024) [reply Untitled.ipynb:1]: Sure!

File: qualtran/bloqs/qsp/.fft_qsp_error_analysis.ipynb.swp
Comment (anurudhp) [inline qualtran/bloqs/qsp/.fft_qsp_error_analysis.ipynb.swp:1]: please delete this file

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+          P: Co-efficients of a complex QSP polynomial
+          num_modes: The number of modes used in the FFT operation. The more modes, the more accurate the result.
+          tolerance: The maximum allowable amount that the sum of the squares can be off from the unit circle. This is
+            mainly used for scaling and does not directly affect the speed of the calculation.
+
+    Returns:
+        The complementary polynomial, Q.
+
+    References:
+    [Complementary polynomials in quantum signal processing](https://arxiv.org/abs/2406.04246)
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:41]:
```suggestion
        [Complementary polynomials in quantum signal processing](https://arxiv.org/abs/2406.04246)
```

File: qualtran/bloqs/qsp/fft_qsp.py [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+    def calculate_coeff(poly: np.ndarray) -> np.ndarray:
+        """Compute coefficients of Q
+
+        Calculates the coefficients and truncates them to the proper degree.
+
+        Args:
+            poly: The input polynomial, P.
+        Returns:
+            The complementary polynomial, Q.
+        """
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp.py:85]: please update this docstring to explain the function, i.e., fft of get modes.

You can just write a one line explanation, no need for Args and returns etc.

File: qualtran/bloqs/qsp/fft_qsp_test.py [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+    random_state = np.random.RandomState(102)
+    num_modes = 500
+    tolerance = 1e-6
+    for _ in range(5):
+        P = random_qsp_polynomial(degree, random_state=random_state)
+        U = MatrixGate.random(bitsize=bitsize, random_state=random_state)
+        Q = fft_complementary_polynomial(P, tolerance=tolerance, num_modes=num_modes)
+        verify_generalized_qsp(U, P, Q=Q, tolerance=tolerance)
+
+
```
Comment (anurudhp) [inline qualtran/bloqs/qsp/fft_qsp_test.py:70]:
```suggestion
```

File: qualtran/bloqs/qsp/fft_qsp.py (hunk 21 above)
Comment (tanujkhattar) [reply qualtran/bloqs/qsp/fft_qsp.py:85]: Having Args and returns is fine I think.

**Pull request merged**